### PR TITLE
Add nightly build using latest gcc and c++23

### PR DESCRIPTION
**Pull request by Rombur**
Add nightly build using gcc 13.1 with C++23 and fix warnings

```
 error: implicit capture of 'this' via '[=]' is deprecated in C++20 [-Werror=deprecated]
```

### CODE REVIEW

**Comment by masterleinad** [comment]
Is there any chance that we get notified when nightly tests fail? Can we at least have a badge or so?
I wasn't aware that `spack-cuda` is failing since 07/12, see https://cloud.cees.ornl.gov/jenkins-ci/job/Kokkos-nightly/.

**Review by masterleinad — APPROVED** [review]

**Comment by dalg24** [comment]
Please point to the place where you tested that configuration

**Comment by Rombur** [comment]
> Please point to the place where you tested that configuration

I tried the configuration on the machine that runs the nightly.

File: .jenkins_nightly [hunk 1]
```diff
@@ -70,6 +70,35 @@ pipeline {
...
+                          apt-get clean && rm -rf /var/lib/apt/lists/*
+
+                          mkdir -p build && cd build && \
+                          cmake \
+                            -DCMAKE_BUILD_TYPE=Release \
+                            -DCMAKE_CXX_STANDARD=23 \
+                            -DCMAKE_CXX_FLAGS=-Werror \
+                            -DKokkos_ARCH_NATIVE=ON \
+                            -DKokkos_ENABLE_COMPILER_WARNINGS=ON \
+                            -DKokkos_ENABLE_TESTS=ON \
```
Comment (cz4rs) [inline .jenkins_nightly:95]: Can you also enable examples and benchmarks?

**Comment by dalg24** [comment]
Waiting for you to answer to Cezary before I merge

**Review by cz4rs — APPROVED** [review]

**Comment by dalg24** [comment]
Did you test that this actually builds and passes?

**Comment by Rombur** [comment]
> Did you test that this actually builds and passes?

I didn't. I thought it would work but I was too optimistic...

**Comment by Rombur** [comment]
Should be good now. I have updated the description of the PR.

**Pull request merged**